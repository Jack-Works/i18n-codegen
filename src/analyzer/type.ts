import type { Diagnostic, SourceFile } from 'typescript'
import { TranslateSymbol } from './symbol.js'
import type { ParserInput } from './parser_input.js'
import type { Project } from './project.js'

export { GeneratorList as EmitterList, ParserList } from '../json-schema.js'
export type Parser = (input: ParserInput<any>, namespace: string) => ParsedFile
export interface ParsedFile {
    filename: string
    source_file: SourceFile
    members: TranslateSymbol[]
    namespaces: TranslateSymbol[]
    diagnostics: Diagnostic[]
}
export type Emitter<Options = object> = (
    parsed_file: ParsedFile,
    options: Options,
    project: Project,
    output_base: string
) => EmittedFile
export interface EmittedFile {
    diagnostics: Diagnostic[]
    /** Map<absolute_output_path, content> */
    files: Map<string, string>
}
export enum DiagnosticCode {
    cannot_open_file_0,
    translation_file_should_contain_an_object,
    non_string_value_is_not_supported_yet,
    invalid_json_content,
    file_0_is_not_in_the_project,
    nested_namespace_or_object_generation_is_not_supported_yet,
    string_0_is_defined_multiple_times_in_namespace_1_this_might_be_caused_by_misconfiguration_of_i18n_codegen_or_if_you_really_mean_it_it_might_cause_problem_in_runtime,
    the_parameter_of_nesting_reference_cannot_be_parsed_as_a_valid_json_therefore_the_type_inference_of_this_string_might_be_incorrect,
    the_parameter_of_nesting_reference_must_be_a_json_object_but_found_0,
}
export enum DiagnosticMessage {
    cannot_open_file_0 = 'Cannot open file $0',
    translation_file_should_contain_an_object = 'Translation file should contain an object',
    non_string_value_is_not_supported_yet = 'Non-string value is not supported yet',
    invalid_json_content = 'Invalid JSON content',
    file_0_is_not_in_the_project = 'File "$0" is not in the project',
    nested_namespace_or_object_generation_is_not_supported_yet = 'Nested namespace or object generation is not supported yet',
    string_0_is_defined_multiple_times_in_namespace_1_this_might_be_caused_by_misconfiguration_of_i18n_codegen_or_if_you_really_mean_it_it_might_cause_problem_in_runtime = 'String "$0" is defined multiple times in namespace "$1". This might be caused by misconfiguration of i18n-codegen or if you really mean it, it might cause problem in runtime.',
    the_parameter_of_nesting_reference_cannot_be_parsed_as_a_valid_json_therefore_the_type_inference_of_this_string_might_be_incorrect = 'The parameter of nesting reference cannot be parsed as a valid JSON, therefore the type inference of this string might be incorrect.',
    the_parameter_of_nesting_reference_must_be_a_json_object_but_found_0 = 'The parameter of nesting reference must be a JSON object but found "$0"',
}
