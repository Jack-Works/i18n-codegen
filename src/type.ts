import { readFileSync } from 'fs'
import { SourceFile } from 'typescript'
import { Position } from './utils/position'
import { JSONNode, parseJson } from './utils/parseJSON'
import { dirname, join, relative } from 'path'
import { GeneratorList, ParserList } from './json-schema'

export { GeneratorList, ParserList } from './json-schema'
export interface PluginConfig {
    input: string
    output: string
    parser: ParserList
    generator: GeneratorList
    generatorOptions?: object
    parserOptions?: object
}
export type Parser = (input: ParserInput<any>) => ParseNode<any>
/** Result is a Map<absolute path, file content> */
export type Generator = (input: GeneratorInput<any, any>) => Map<string, GeneratorResult>
export type GeneratorResult = string
export class ParserInput<Options> {
    private constructor(
        public sourceFile: JSONNode,
        public readonly originalFile: string,
        public readonly parserOptions?: Options,
    ) {}
    readonly mockSourceFile: SourceFile = { text: this.originalFile } as any
    static fromFileSystem<T>(path: string, parserOptions?: T) {
        const orig = readFileSync(path, 'utf-8')
        return new ParserInput(parseJson(path, orig), orig, parserOptions)
    }
}
export class GeneratorInput<R extends object, Options> {
    constructor(
        public parseResult: R,
        public inputPath: string,
        public outputBase: string,
        public generatorOptions?: Options,
    ) {}
    relativeToInput(x: string) {
        return join(this.inputPath, x)
    }
    relativeToOutput(x: string) {
        return join(this.outputBase, x)
    }
    get outputRelativePathToInput() {
        return relative(dirname(this.outputBase), this.inputPath).replace(/\\/g, '/')
    }
}
export type StringParseNode<T extends object> = {
    type: 'key'
    position: Position
    value: string
    value_position: Position
} & T
export type MapParseNode<T extends object> = {
    type: 'object'
    position: Position
    items: Map<string, ParseNode<T>>
}
export type ArrayParseNode<T extends object> = {
    type: 'array'
    position: Position
    items: ParseNode<T>[]
}
export type ParseNode<T extends object> = MapParseNode<T> | ArrayParseNode<T> | StringParseNode<T>
