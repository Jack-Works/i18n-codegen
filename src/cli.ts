#!/usr/bin/env node
import yargs from 'yargs'
// @ts-ignore
import { hideBin } from 'yargs/helpers'
import { runCli } from './index'

yargs(hideBin(process.argv))
    .scriptName('i18n-codegen')
    .option('config', {
        alias: 'c',
        type: 'string',
        description: 'Specify a different configuration file',
        default: './.i18n-codegen.json',
    })
    .option('cwd', {
        type: 'string',
        description: 'Current working directory',
    })
    .option('watch', {
        alias: 'w',
        type: 'string',
        description: 'Current working directory',
    })
    .command(
        'gen',
        'Generate the i18n files',
        (yargs) => {
            yargs.positional('watch', {
                describe: 'watch mode',
                default: false,
                alias: 'w',
            })
        },
        runCli,
    ).argv
