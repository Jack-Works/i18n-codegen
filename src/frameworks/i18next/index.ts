export { i18next_react_hooks_emitter } from './emitter/react-hooks.js'
export { i18next_parser } from './parser/index.js'
