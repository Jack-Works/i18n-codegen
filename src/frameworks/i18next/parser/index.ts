export { i18NextParser } from './i18n-next.js'
