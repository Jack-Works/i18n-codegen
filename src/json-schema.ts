/**
 * This file must not contains any complex TypeScript knowledge that cannot be understand by https://app.quicktype.io/
 * We use this tool to generate JSON Schema for the config file.
 *
 * Don't forget to add these lines under the JSON:
    "type": "object",
    "$ref": "#/definitions/ConfigFile"
 */

/** #TopLevel */
/** Configuration format of i18n-codegen */
export type ConfigFile = { version: 1; list: Config[] }
/**
 * Each configuration entry
 */
export interface Config {
    /** The input file, must be a JSON (can be a JSON with comment) */
    input: string
    /** The base name of the output files. It might generate multiple files. */
    output: string
    parser: ParserList | ({ type: ParserList.i18next } & Parser_I18NextConfig)
    generator: GeneratorList | ({ type: GeneratorList.i18next_reactHooks } & Generator_I18Next_ReactHooks)
}
export enum ParserList {
    i18next = 'i18next',
}
export enum GeneratorList {
    i18next_reactHooks = 'i18next/react-hooks',
}
//#region i18next
export interface Parser_I18NextConfig {
    nsSeparator?: string
    pluralSeparator?: string
    contextSeparator?: string
}
export interface Generator_I18Next_ReactHooks {
    /** The namespace of this generator should use */
    namespace?: string
    /** The hooks name, must starts with "use" */
    hooks?: string
    /** The component name that provides typed version of Trans component. */
    trans?: string
    /**
     * Generate sourcemap for .d.ts file
     * @default true
     */
    sourceMap?: boolean | 'inline'
    /**
     * Use ES6 Proxy for .js file
     * @default true
     */
    es6Proxy?: boolean
}
//#endregion
