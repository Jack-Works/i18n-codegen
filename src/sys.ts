import { readFileSync as r, writeFileSync as w } from 'fs'
import type { EmittedFile } from './analyzer/type.js'
import type { Diagnostic } from 'typescript'
import { isAbsolute } from 'path'
export const sys = {
    readFile(path: string) {
        try {
            return r(path, 'utf-8')
        } catch (error) {
            return new Error(`Failed to read file ${path}`, { cause: error })
        }
    },
    writeFile(path: string, str: string) {
        return w(path, str, 'utf-8')
    },
}

export function write_emitted_result(result: EmittedFile, error_reporter: (diagnostics: Diagnostic[]) => void) {
    if (result.diagnostics.length) error_reporter(result.diagnostics)
    for (const [file, content] of result.files) {
        if (!isAbsolute(file))
            throw new Error(`Internal error: emitted file should be an absolute path but found ${file}`)
        sys.writeFile(file, content)
    }
}
