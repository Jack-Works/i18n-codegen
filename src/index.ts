import { i18NextParser } from './frameworks/i18next/parser/i18n-next'
import { Config, ConfigFile, GeneratorList, ParserList } from './json-schema'
import { GeneratorInput, ParserInput } from './type'
import { dirname, resolve } from 'path'
import { i18next_reactHooksGenerator } from './frameworks/i18next'
import { readFileSync, writeFileSync } from 'fs'
import { watch } from 'chokidar'
export * from './json-schema'

type Args = [config: Pick<Config, 'generator' | 'parser'>, absoluteInputPath: string, absoluteOutputPath: string]
type E = (error: unknown, ...args: Args) => void
export function handleConfig(...[config, input, output]: Args) {
    let { generator, parser } = config
    if (typeof generator === 'string') generator = { type: generator }
    if (typeof parser === 'string') parser = { type: parser }

    const table = { [ParserList.i18next]: i18NextParser }
    const table2 = { [GeneratorList.i18next_reactHooks]: i18next_reactHooksGenerator }

    const result = table[parser.type](ParserInput.fromFileSystem(input, parser))
    const files = table2[generator.type](new GeneratorInput(result, input, output, generator))

    for (const [file, content] of files) {
        writeFileSync(file, content)
    }
}

const watchOptions = { atomic: true, awaitWriteFinish: true, ignoreInitial: false }
export function watchConfig(onError: E, ...[config, input, output]: Args) {
    const watcher = watch(input, watchOptions)
    watcher.on('all', () => {
        try {
            handleConfig(config, input, output)
        } catch (e) {
            onError(e, config, input, output)
        }
    })
    return () => watcher.close()
}

export function runConfigList(watchMode: true, onError: E, configs: Args[]): () => void
export function runConfigList(watchMode: false, onError: E, configs: Args[]): void
export function runConfigList(watchMode: boolean, onError: E, configs: Args[]): void | (() => void)
export function runConfigList(watchMode: boolean, onError: E, configs: Args[]) {
    if (watchMode) {
        const cancel = configs.map((x) => watchConfig(onError, ...x))
        return () => cancel.forEach((x) => x())
    }
    for (const x of configs) {
        try {
            handleConfig(...x)
        } catch (e) {
            onError(e, ...x)
        }
    }
    return
}
export function runConfigFile(watchMode: true, onRecoverableError: E, configFilePath: string): () => void
export function runConfigFile(watchMode: false, onRecoverableError: E, configFilePath: string): void
export function runConfigFile(watchMode: boolean, onRecoverableError: E, configFilePath: string): void | (() => void)
export function runConfigFile(watchMode: boolean, onRecoverableError: E, configFilePath: string) {
    const config: ConfigFile = JSON.parse(readFileSync(configFilePath, 'utf-8'))
    if (config.version !== 1) return
    const base = dirname(configFilePath)
    const configList = config.list.map(({ input, output, ...rest }) => {
        const absoluteInputPath = resolve(base, input)
        const absoluteOutputPath = resolve(base, output)
        return [rest, absoluteInputPath, absoluteOutputPath] as Args
    })
    return runConfigList(watchMode, onRecoverableError, configList)
}
export function runCli(argv: { config?: string; cwd?: string; watch?: boolean }, onError: E) {
    const config = resolve(process.cwd(), argv.config || './.i18n-codegen.json')
    if (argv.watch) {
        const watcher = watch(config, watchOptions)
        let r = runConfigFile(true, onError, config)
        watcher.addListener('all', () => {
            r()
            r = runConfigFile(true, onError, config)
        })
        return () => watcher.close()
    } else return runConfigFile(false, onError, config)
}
