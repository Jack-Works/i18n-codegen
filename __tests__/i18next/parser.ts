import { i18NextParser } from '../../src/frameworks/i18next'
import { it, expect } from '@jest/globals'

import { join } from 'path'
import { ParserInput } from '../../src/type'

it('should parse i18next format correctly', () => {
    expect(i18NextParser(ParserInput.fromFileSystem(join(__dirname, './example.json')))).toMatchSnapshot(
        'i18next-parse-example',
    )
})
