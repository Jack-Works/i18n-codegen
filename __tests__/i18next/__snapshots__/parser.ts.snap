// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse i18next format correctly: i18next-parse-example 1`] = `
Object {
  "plurals": Map {
    "key" => Map {
      "zero" => Object {
        "interpolations": Map {},
        "position": Array [
          13,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "zero",
        "value_position": Array [
          13,
          16,
        ],
      },
      "one" => Object {
        "interpolations": Map {},
        "position": Array [
          14,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "singular",
        "value_position": Array [
          14,
          15,
        ],
      },
      "two" => Object {
        "interpolations": Map {},
        "position": Array [
          15,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "two",
        "value_position": Array [
          15,
          15,
        ],
      },
      "few" => Object {
        "interpolations": Map {},
        "position": Array [
          16,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "few",
        "value_position": Array [
          16,
          15,
        ],
      },
      "many" => Object {
        "interpolations": Map {},
        "position": Array [
          17,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "many",
        "value_position": Array [
          17,
          16,
        ],
      },
      "other" => Object {
        "interpolations": Map {
          "things" => Array [
            Object {
              "content": "things",
              "position": Array [
                18,
                26,
              ],
            },
            "string",
          ],
        },
        "position": Array [
          18,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "other {{things}}",
        "value_position": Array [
          18,
          17,
        ],
      },
    },
  },
  "root": Object {
    "items": Map {
      "normal_key" => Object {
        "interpolations": Map {},
        "position": Array [
          3,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "this is a normal key",
        "value_position": Array [
          3,
          18,
        ],
      },
      "with_param" => Object {
        "interpolations": Map {
          "name" => Array [
            Object {
              "content": "name",
              "position": Array [
                8,
                28,
              ],
            },
            "string",
          ],
        },
        "position": Array [
          8,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "Hello, {{name}}!",
        "value_position": Array [
          8,
          18,
        ],
      },
      "with_tag" => Object {
        "interpolations": Map {},
        "position": Array [
          9,
          4,
        ],
        "tags": Map {
          "i" => Array [
            Object {
              "content": "i",
              "position": Array [
                9,
                18,
              ],
            },
            "JSX.Element",
          ],
        },
        "type": "key",
        "value": "<i>hi</i>",
        "value_position": Array [
          9,
          16,
        ],
      },
      "with_prop_access" => Object {
        "interpolations": Map {
          "author" => Array [
            Object {
              "content": "author.name.last",
              "position": Array [
                10,
                54,
              ],
            },
            "object",
          ],
        },
        "position": Array [
          10,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "I am {{author.name.first}} {{author.name.last}}",
        "value_position": Array [
          10,
          24,
        ],
      },
      "unescaped" => Object {
        "interpolations": Map {
          "var" => Array [
            Object {
              "content": "var",
              "position": Array [
                11,
                31,
              ],
            },
            "string",
          ],
        },
        "position": Array [
          11,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "dangerous {{- var}}",
        "value_position": Array [
          11,
          17,
        ],
      },
      "formatted" => Object {
        "interpolations": Map {
          "date" => Array [
            Object {
              "content": "date, MM/DD/YYYY",
              "position": Array [
                12,
                40,
              ],
            },
            "string",
          ],
        },
        "position": Array [
          12,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "The current date is {{date, MM/DD/YYYY}}",
        "value_position": Array [
          12,
          17,
        ],
      },
      "key$zero" => Object {
        "interpolations": Map {},
        "position": Array [
          13,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "zero",
        "value_position": Array [
          13,
          16,
        ],
      },
      "key$one" => Object {
        "interpolations": Map {},
        "position": Array [
          14,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "singular",
        "value_position": Array [
          14,
          15,
        ],
      },
      "key$two" => Object {
        "interpolations": Map {},
        "position": Array [
          15,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "two",
        "value_position": Array [
          15,
          15,
        ],
      },
      "key$few" => Object {
        "interpolations": Map {},
        "position": Array [
          16,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "few",
        "value_position": Array [
          16,
          15,
        ],
      },
      "key$many" => Object {
        "interpolations": Map {},
        "position": Array [
          17,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "many",
        "value_position": Array [
          17,
          16,
        ],
      },
      "key$other" => Object {
        "interpolations": Map {
          "things" => Array [
            Object {
              "content": "things",
              "position": Array [
                18,
                26,
              ],
            },
            "string",
          ],
        },
        "position": Array [
          18,
          4,
        ],
        "tags": Map {},
        "type": "key",
        "value": "other {{things}}",
        "value_position": Array [
          18,
          17,
        ],
      },
    },
    "position": Array [
      0,
      0,
    ],
    "type": "object",
  },
}
`;
